pkg/api/handlers/oauth/client.go

```diff
@@ -0,0 +1,228 @@
+		return err
+	}
+
+	clientSecret, registrationToken, err := updateClientIfNecessary(req.Context(), req.Storage, &oauthClient)
+	if err != nil {
+		return fmt.Errorf("failed to update client secret: %w", err)
+	}
+
+	return req.Write(convertClient(oauthClient, h.baseURL, clientSecret, registrationToken))
+}
+
+func (h *handler) updateClient(req api.Context) error {
+	var oauthClientManifest types.OAuthClientManifest
+	if err := req.Read(&oauthClientManifest); err != nil {
+		return types.NewErrBadRequest("invalid request body: %v", err)
+	}
+
+	namespace, name, ok := strings.Cut(req.PathValue("client"), ":")
+	if !ok {
+		return types.NewErrBadRequest("invalid client name: %s", name)
+	}
+
+	var oauthClient v1.OAuthClient
+	if err := req.Storage.Get(req.Context(), kclient.ObjectKey{Namespace: namespace, Name: name}, &oauthClient); err != nil {
+		return err
+	}
+
+	oauthClient.Spec.Manifest = oauthClientManifest
+
+	if err := h.validateClientConfig(&oauthClient); err != nil {
+		return err
+	}
+
+	clientSecret, registrationToken, err := updateClientIfNecessary(req.Context(), req.Storage, &oauthClient)
+	if err != nil {
+		return fmt.Errorf("failed to update client secret: %w", err)
+	}
+
+	if err = req.Update(&oauthClient); err != nil {
+		return err
+	}
+
+	return req.Write(convertClient(oauthClient, h.baseURL, clientSecret, registrationToken))
+}
+
+func (h *handler) deleteClient(req api.Context) error {
+	namespace, name, ok := strings.Cut(req.PathValue("client"), ":")
+	if !ok {
+		return types.NewErrBadRequest("invalid client name: %s", name)
+	}
+
+	return req.Delete(&v1.OAuthClient{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+	})
+}
+
+func updateClientIfNecessary(ctx context.Context, c kclient.Client, oauthClient *v1.OAuthClient) (string, string, error) {
+	clientSecret, registrationToken, err := ensureTokenAndSecret(oauthClient)
+	if err != nil {
+		return "", "", err
+	}
+
+	if clientSecret != "" || registrationToken != "" {
+		if err = c.Update(ctx, oauthClient); err != nil {
+			return "", "", err
+		}
+	}
+
+	return clientSecret, registrationToken, nil
+}
+
+func ensureTokenAndSecret(oauthClient *v1.OAuthClient) (string, string, error) {
+	var (
+		clientSecret, registrationToken string
+		err                             error
+	)
+	if oauthClient.Spec.ClientSecretIssuedAt.IsZero() || oauthClient.Spec.ClientSecretExpiresAt.Sub(oauthClient.Spec.ClientSecretIssuedAt.Time)/2 > time.Until(oauthClient.Spec.ClientSecretExpiresAt.Time) {
+		// If the client secret is half-way through its lifetime, then update it.
+		clientSecret = rand.Text() + rand.Text()
+		oauthClient.Spec.ClientSecretHash, err = bcrypt.GenerateFromPassword([]byte(clientSecret), bcrypt.DefaultCost)
+		if err != nil {
+			return "", "", err
+		}
+
+		oauthClient.Spec.ClientSecretIssuedAt = metav1.NewTime(time.Now())
+		oauthClient.Spec.ClientSecretExpiresAt = metav1.NewTime(time.Now().Add(time.Hour + 15*time.Minute))
```

Comment:
is 15 minute too short for client secret to expire? seems like when access token expire the client will have to register to refresh their client secret first in order to refresh access token.

Reply:
It expires in an hour and 15 minutes, so that the client can refresh a token. Let me know if that changes your opinion.